src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs

```diff
@@ -190,15 +217,66 @@ public List<string> ExecuteCompareExchangeCommands(ClusterOperationContext conte
 
             return null;
         }
-        
+
+        private void EnsureAtomicDocumentWrites(DatabaseTopology dbTopology, ClusterOperationContext context, Table items, long index)
+        {
+            if (SerializedDatabaseCommands == null)
+                return;
+            if(SerializedDatabaseCommands.TryGet(nameof(DatabaseCommands), out BlittableJsonReaderArray commands) == false)
+                return;
+            
+            ClusterCommands ??= new List<ClusterTransactionDataCommand>();
+            foreach (BlittableJsonReaderObject dbCmd in commands)
+            {
+                var cmdType = dbCmd[nameof(ClusterTransactionDataCommand.Type)].ToString();
+                var docId = dbCmd[nameof(ClusterTransactionDataCommand.Id)].ToString();
+                var atomicGuardKey = GetAtomicGuardKey(docId);
+                var changeVector = dbCmd[nameof(ClusterTransactionDataCommand.ChangeVector)]?.ToString();
+                long changeVectorIndex = -1;
+                if (changeVector != null)
+                    changeVectorIndex = ChangeVectorUtils.GetEtagById(changeVector, dbTopology.ClusterTransactionIdBase64);
+                ClusterCommands.Add(new ClusterTransactionDataCommand
+                {
+                    Type = cmdType switch {
+                    
+                        nameof(CommandType.PUT) =>CommandType.CompareExchangePUT,
+                        nameof(CommandType.DELETE) => CommandType.CompareExchangeDELETE,
+                        _ => throw new ArgumentOutOfRangeException()   
+                    },
+                    Id = atomicGuardKey,
+                    Index = changeVectorIndex,
+                    Document = context.ReadObject(new DynamicJsonValue { ["Id"] = docId }, "cmp-xchg-content"),
+                    Error = $"Guard compare exchange value '{atomicGuardKey}' index does not match the transaction index's {changeVectorIndex} change vector on {docId}"
```

Comment:
Do we have to send this error already with the command data? Cannot we just recognize that somehow after the validation fails and then add / expose this error when needed?

Reply:
The problem is that we don't have a good place to give the _context_ for the error.
The actual validation happens later, and we may also have other compare exchange operations explicitly modified.

Reply:
My concern was that we will be _always_ sending this error string while we already have everything - `atomicGuardKey`, `changeVectorIndex` and `docId` - inside that command to build that error message.